### ISSUE TITLE
Bilinear function is updated wrongly only with GPU.

### ISSUE THREAD

**Issue by suzukikbp** [chainer/chainer#1888]
I have the setting below. This works correctly with CPU, though it returns NAN immediately when I use single GPU. I can use GPU without any problem when I run a model without Bilinear function, so I am guessing the Bilinear function with GPU is somewhat wrong. From what I have seen, ge1 and ge2 in the backward part of BilinearFunction always return 0 with GPU (gW doesnt). Thanks in advance!

    class Settings():
        def __init__(self) :
            cls = 10
            # source1
            self.cv1_i = 3, 20, 5]
            self.cv2_i = [self.cv1_i[1], 50, 5]
            self.fc1_i = [800, 500] 
            self.fc2_i = [self.fc1_i[1], 50] 

            # source2
            self.cv1_w = [1, 20, (5,1)]
            self.cv2_w = [self.cv1_w[1], 50, (5,1)]
            self.fc1_w = [3050, 500]
            self.fc2_w = [self.fc1_w[1], 50]

            # fusion
            self.fc3 = [self.fc1_i[1], self.fc1_w[1], 50]
            self.fc4 = [self.fc3[2], cls]

    class CNN_ful(chainer.Chain):
        def __init__(self, s):
            super(CNN_ful, self).__init__(
                # source1
                conv1_i=L.Convolution2D(s.cv1_i[0], s.cv1_i[1], s.cv1_i[2]),
                bn1_i=L.BatchNormalization(s.cv1_i[1]),
                conv2_i=L.Convolution2D(s.cv2_i[0], s.cv2_i[1], s.cv2_i[2]),
                bn2_i=L.BatchNormalization(s.cv2_i[1]),
                fc1_i=L.Linear(s.fc1_i[0], s.fc1_i[1]),
                bn3_i=L.BatchNormalization(s.fc1_i[1]),
                # source2
                conv1_w=L.Convolution2D(s.cv1_w[0], s.cv1_w[1], s.cv1_w[2]),
                bn1_w=L.BatchNormalization(s.cv1_w[1]),
                conv2_w=L.Convolution2D(s.cv2_w[0], s.cv2_w[1], s.cv2_w[2]),
                bn2_w=L.BatchNormalization(s.cv2_w[1]),
                fc1_w=L.Linear(s.fc1_w[0], s.fc1_w[1]),
                bn3_w=L.BatchNormalization(s.fc1_w[1]),
                # fusion
                fc3=L.Bilinear(s.fc3[0], s.fc3[1], s.fc3[2], nobias=True),
                bn3=L.BatchNormalization(s.fc3[2]),
                fc4=L.Linear(s.fc4[0], s.fc4[1])
            )
            self._train = True

        @property
        def train(self):
            return self._train

        @train.setter
        def train(self, value):
            self._train = value

        def __call__(self, xi, xw, t):
            # source1
            test = not self.train
            x = xi
            h = self.bn1_i(F.relu(self.conv1_i(x)), test=test)
            h = F.max_pooling_2d(h, 2)
            h = self.bn2_i(F.relu(self.conv2_i(h)), test=test)
            h = F.max_pooling_2d(h, 2)
            hi = F.dropout(self.bn3_i(F.relu(self.fc1_i(h)), test=test), train=self.train)
            # source2
            x = xw
            h = self.bn1_w(F.relu(self.conv1_w(x)), test=test)
            h = F.max_pooling_2d(h, 2)
            h = self.bn2_w(F.relu(self.conv2_w(h)), test=test)
            h = F.max_pooling_2d(h, 2)
            hw = F.dropout(self.bn3_w(F.relu(self.fc1_w(h)), test=test), train=self.train)
            # fusion
            h = F.dropout(self.bn3(F.relu(self.fc3(hi, hw)),test=test), train=self.train)
            self.y = self.fc4(h)
            
            loss = F.softmax_cross_entropy(self.y, t)
            return loss

**Comment by delta2323** [comment]
Hi, thank you for reporting the problem.

I used your model with a slight modification explained later and found that the gradient of the weight matrix of `fc4` is already 0 in GPU case. Could you check the gradient in your environment, too?

I remove dropout to remove stochasticity and change `s.fc1_w[0]` from 3050 to `None` because I could not infer the input shape of `xw`. After that, I ran the following code.
```
s = Settings()
m_cpu = CNN_ful(s)
m_gpu = copy.deepcopy(m_cpu).to_gpu()

x = chainer.Variable(np.random.uniform(0, 1, (2, 3, 25, 25)).astype(np.float32))
y = chainer.Variable(np.random.uniform(0, 1, (2, 1, 37, 37)).astype(np.float32))
t = chainer.Variable(np.random.randint(0, 10, (2,)).astype(np.int32))


# cpu                                                                                                                                                                                                                                                                                    
m_cpu.cleargrads()
l = m_cpu(x, y, t)
l.backward()

l_cpu = copy.deepcopy(l.data)
gx_cpu = copy.deepcopy(x.grad)
gy_cpu = copy.deepcopy(y.grad)
g_cpu = copy.deepcopy(m_cpu.fc4.W.grad)


# gpu                                                                                                                                                                                                                                                                                    
x.to_gpu()
y.to_gpu()
t.to_gpu()   
x.cleargrad()
y.cleargrad()
t.cleargrad()
m_gpu.cleargrads()
l = m_gpu(x, y, t)
l.backward()

l_gpu = copy.deepcopy(l.data.get())
gx_gpu = copy.deepcopy(x.grad.get())
gy_gpu = copy.deepcopy(y.grad.get())
g_gpu = copy.deepcopy(m_gpu.fc4.W.grad.get())


# test                                                                                                                                                                                                                                                                                   
def error(x, y):
    return np.linalg.norm(x - y) / np.linalg.norm(x)

print(error(l_cpu, l_gpu))  # => 7.65248e-05
print(error(g_cpu, g_gpu))   # => 1.0

print(np.linalg.norm(g_cpu), np.linalg_norm(g_gpu))  # => 0.00206023 0.0
```

**Comment by suzukikbp** [comment]
Hi, Thank you for replying!
Actually, your code works fine for me as below under the modifications you mentioned. Though, I would like to know how I can check these values after 10 iterations since I clearly see the diference between CPU and GPU after 10 iterations with my original dataset. 

    0.00015263
    1.23028
    (0.018134331, 0.013040739) # fc3
    (0.0023767757, 0.0012177534) # fc4

BTW I identify my environment just in case.
    chainer: 1.17.0
    Python: 2.7.6
    cuda: 7.5
    cuDNN: 4.0.7
    GPU: GeForce TITAN X
    OS: Ubuntu 14.04.5 LTS

Thanks!

**Comment by delta2323** [comment]
> I would like to know how I can check these values after 10 iterations

If you use trainer, one of the way is to make a custom extension of `Trainer`.
